Helper/Service.php

```diff
@@ -0,0 +1,354 @@
+        $captureRequest->setRequestResource($transactionObject);
+
+        /** @var ResponseServiceInterface $responseService */
+        $responseService = $captureRequest->send();
+
+        $response = $responseService->getResponseData();
+
+        $this->logger->debug(json_encode($response));
+
+        return $responseService;
+    }
+
+    /**
+     * @param string $paymentUri
+     * @return OrderInterface|QuoteInterface|bool
+     * @throws ClientException
+     * @throws ServiceException
+     * @throws NoSuchEntityException
+     *
+     * @SuppressWarnings(PHPMD.LongVariable)
+     */
+    public function getPaymentData($paymentUri)
+    {
+        /**
+         * $paymentParams[0] Payment URI
+         * $paymentParams[1] Payment Instrument
+         * $paymentParams[2] Payment ID
+         */
+        $validUri = preg_match('|/psp/([^/]+)/payments/([^/]+)|', $paymentUri, $paymentParams);
+        if (!$validUri) {
+            return false;
+        }
+
+        list($paymentUri, $instrument, $paymentId) = $paymentParams;
+
+        $instrument = $this->getInstrument($instrument);
+
+        $paymentData = $this->paymentDataHelper->getByPaymentId($paymentId);
+
+        /** @var GetCurrentPayment $serviceRequest */
+        $serviceRequest = $this->requestService->init(ucfirst($instrument), 'GetPayment');
+        $serviceRequest->setRequestEndpoint($paymentUri);
+
+        /** @var GetCurrentPaymentInterface $currentPayment */
+        $currentPaymentResponse = $serviceRequest->send();
+        $currentPayment = $currentPaymentResponse->getResponseResource();
+
+        $this->logger->debug(
+            sprintf(
+                'Current Payment Response: \'%s\'',
+                json_encode($currentPaymentResponse->getResponseData())
+            )
+        );
+
+        $paymentData->setIntent($currentPayment->getPayment()->getIntent());
+        $paymentData->setState($currentPayment->getPayment()->getState());
+        $paymentData->setAmount($currentPayment->getPayment()->getAmount());
+        $this->paymentDataHelper->update($paymentData);
+
+        return $paymentData;
+    }
+
+    /**
+     * @param string $transactionUri
+     * @param string $paymentUri
+     * @return TransactionInterface|false
+     * @throws ClientException
+     * @throws NoSuchEntityException
+     * @throws ServiceException
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     * @SuppressWarnings(PHPMD.UnusedLocalVariable)
+     */
+    public function getTransactionData($transactionUri, $paymentUri)
+    {
+        /**
+         * $transactionParams[0] Transaction URI
+         * $transactionParams[1] Payment Instrument
+         * $transactionParams[2] Payment Resource ID
+         * $transactionParams[3] Transaction Type
+         * $transactionParams[4] Transaction ID
+         */
+        $validUri = preg_match('|/psp/([^/]+)/payments/([^/]+)/([^/]+)/([^/]+)|', $transactionUri, $transactionParams);
```

Comment:
No URI parsing, please. If you want to check for a valid URI, just use `filter_var($url, FILTER_VALIDATE_URL)` or something similar. The exact structure of the URI should be of no interest to the Magento plugin.